Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/split/CheckInSplitter.kt

```diff
@@ -0,0 +1,94 @@
+package de.rki.coronawarnapp.eventregistration.checkins.split
+
+import de.rki.coronawarnapp.eventregistration.checkins.CheckIn
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.seconds
+import org.joda.time.Instant
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+import kotlin.math.ceil
+import kotlin.math.max
+
+/**
+ * Splits a [CheckIn] by midnight UTC time into multiple [CheckIn]s across multiple days.
+ *
+ * @return [List] of [CheckIn]s
+ */
+fun CheckIn.splitByMidnightUTC(): List<CheckIn> {
+    val startTimeSeconds = checkInStart.seconds
+    val endTimeSeconds = checkInEnd?.seconds ?: 0L
+    val durationSeconds = endTimeSeconds - startTimeSeconds
+
+    Timber.i("durationSeconds=$durationSeconds")
+
+    val durationDays = max(1L, ceil(durationSeconds.toDouble() / DAY_IN_SECONDS).toLong())
+    Timber.i("durationDays=$durationDays")
+
+    return Array(durationDays.toInt()) { day ->
+        checkInCopy(
+            day.toLong(),
+            durationDays,
+            startTimeSeconds,
+            endTimeSeconds
+        )
+    }.toList().also { it.print() }
+}
+
+private fun CheckIn.checkInCopy(
+    day: Long,
+    durationDays: Long,
+    startTimeSeconds: Long,
+    endTimeSeconds: Long
+): CheckIn = when {
+    isFirstDay(day) && !isLastDay(day, durationDays) ->
+        copy(
+            checkInEnd = toInstant(
+                seconds = toMidnightUTC(startTimeSeconds) + daysInSeconds(day + 1)
+            )
+        )
+
+    !isFirstDay(day) && isLastDay(day, durationDays) ->
+        copy(
+            checkInStart = toInstant(seconds = toMidnightUTC(endTimeSeconds))
+        )
+
+    !isFirstDay(day) && !isLastDay(day, durationDays) ->
+        copy(
+            checkInStart = toInstant(
+                seconds = toMidnightUTC(startTimeSeconds) + daysInSeconds(day)
+            ),
+            checkInEnd = toInstant(
+                seconds = toMidnightUTC(startTimeSeconds) + daysInSeconds(day + 1)
+            )
+        )
+
+    else -> copy()
+}
+
+private fun toMidnightUTC(timestampSeconds: Long): Long =
```

Comment:
could be extension function.
also below for **daysInnSeconds**, **toInstant**, **isFirstDay** etc

Reply:
For `isFirstDay` ,` isLastDay`, `toMidnightUTC`  it is nice to have them as extensions. For 
`toInstant` and `daysInSeconds` , it does not look simpler. I keep as it is  

```kotlin
// Extensions 
(day + 1).daysInSeconds()
(startTimeSeconds.toMidnightUTC() + daysInSeconds(day + 1)).toInstant()

// Normal functions
 daysInSeconds(day + 1)
toInstant(
  seconds = startTimeSeconds.toMidnightUTC() + daysInSeconds(day)
)
```